apps/fz_http/lib/fz_http/devices.ex

```diff
@@ -63,24 +63,17 @@ defmodule FzHttp.Devices do
   def get_device!(id), do: Repo.get!(Device, id)
 
   def create_device(attrs \\ %{}) do
-    # XXX: insert sometimes fails with deadlock errors, probably because
-    # of the giant SELECT in queries/inet.ex. Find a way to do this more gracefully.
-    {:ok, result} =
-      Repo.transaction(fn ->
-        %Device{}
-        |> Device.create_changeset(attrs)
-        |> Repo.insert()
-      end)
-
-    case result do
-      {:ok, _device} ->
+    attrs
+    |> Device.create_changeset()
+    |> Repo.insert()
```

Comment:
Oh, I did not see the IPvX queries. But..

I don't think wrapping the query in a `Repo.transaction/1` did actually change anything. Postgres uses a read-committed isolation level, which means that two separate database transactions can execute two queries and have the same result, and then still have a conflict during the `Repo.insert/1`. I also noticed the comments there that next IP generation should be an insert, which also won't solve the problem.

I like the SERIAL solution, to update it you don't need a migration, it's as simple as `SELECT setval('device_ipv4_seq', ?, true);`.

Another way would be to match on function return value and if it's a changeset with a conflict on ipv4/6 fields we can simply retry the function execution, it would be less optimal than a serial though but will be very simple to implement.

Reply:
I suggest that we remove the transaction wrap now (as it should not do anything anyways) and create an issue to fix the underlying issue there and make tests async, I can take that a bit later.

Reply:
👍

Reply:
Ah great, didn't realize `SERIAL` was that flexible. That's awesome.